maildir.go

```diff
@@ -203,14 +208,32 @@ func (d Dir) Filename(key string) (string, error) {
 			return guess, nil
 		}
 	}
-	matches, err := filepath.Glob(filepath.Join(string(d), "cur", key+"*"))
+	file, err := os.Open(filepath.Join(string(d), "cur"))
 	if err != nil {
 		return "", err
 	}
-	if n := len(matches); n != 1 {
-		return "", &KeyError{key, n}
+	defer file.Close()
+
+	// search for a valid candidate (in blocks of readdirChunk)
+	for {
+		var names []string
+		names, err = file.Readdirnames(readdirChunk)
```

Comment:
`Readdirnames` will never return a non-nil slice *and* an error. The code can be simplified a bit:

```go
names, err := file.Readdirnames(readdirChunk)
if errors.Is(err, io.EOF) {
    return "", &KeyError{key, 0}
} else if err != nil {
    return "", err
}
```